public/ocr/ocr_main.js

```diff
@@ -844,6 +863,291 @@ async function resetDevices() {
 
 navigator.mediaDevices.addEventListener('devicechange', resetDevices);
 
+let start_time;
+let frame_duration;
+const EVERDRIVE_N8_PRO = { usbVendorId: 0x483, usbProductId: 0x5740 };
+let everdrive;
+let everdrive_reader;
+let everdrive_writer;
+const EVERDRIVE_CMD_SEND_STATS = 0x42;
+const EVERDRIVE_CMD_MEM_WR = 0x1a;
+const EVERDRIVE_ADDR_FIFO = 0x1810000;
+const GAME_FRAME_SIZE = 232; // 0xe8
+const PIECE_ORIENTATION_TO_PIECE_ID = [
+	0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5, 5, 5, 5, 6, 6,
+];
+const TILE_ID_TO_NTC_BLOCK_ID = new Map();
+TILE_ID_TO_NTC_BLOCK_ID.set(0xef, 0);
+TILE_ID_TO_NTC_BLOCK_ID.set(0x7b, 1);
+TILE_ID_TO_NTC_BLOCK_ID.set(0x7d, 2);
+TILE_ID_TO_NTC_BLOCK_ID.set(0x7c, 3);
+
+let data_frame_buffer = new Uint8Array(GAME_FRAME_SIZE);
+
+async function initCaptureFromEverdrive() {
+	frame_duration = 1000 / config.frame_rate;
+	const ports = await navigator.serial.getPorts();
+	if (ports.length) {
+		everdrive = ports.find(port => {
+			const { usbProductId, usbVendorId } = port.getInfo();
+			return (
+				usbVendorId === EVERDRIVE_N8_PRO.usbVendorId &&
+				usbProductId === EVERDRIVE_N8_PRO.usbProductId
+			);
+		});
+
+		if (everdrive) {
+			captureFromEverdrive();
+			return;
+		}
+	}
+
+	everdrive = await navigator.serial.requestPort({
+		filters: [EVERDRIVE_N8_PRO],
+	});
+
+	if (everdrive) {
+		captureFromEverdrive();
+	}
+}
+
+async function captureFromEverdrive() {
+	try {
+		await everdrive.open({ baudRate: 115200, bufferSize: GAME_FRAME_SIZE }); // plenty of speed for 60fps data frame from gym are 132 bytes: 132x60=7920
+	} catch (err) {
+		console.warn(err);
+		// assume port is already open for now
+		// TODO: better error checking
+	}
+
+	everdrive_reader = everdrive.readable.getReader({ mode: 'byob' });
+	everdrive_writer = everdrive.writable.getWriter();
+
+	// verify we have a real everdrive by sending a GET_STATUS command (expecting [0x00, 0xA5] as response)
+	const bytes = [
+		'+'.charCodeAt(0),
+		'+'.charCodeAt(0) ^ 0xff,
+		0x10, // getStatus
+		0x10 ^ 0xff,
+	];
+
+	await everdrive_writer.write(new Uint8Array(bytes));
+	const { value, done } = await everdrive_reader.read(data_frame_buffer);
+
+	if (value[0] !== 0 || value[1] !== 0xa5) {
```

Comment:
This can trigger erroneously if the fifo buffer still has data left in it.  I was thinking maybe checking the last two bytes instead of the first two?  Additionally, the gym hack can be modified to include the same expected bytes at the tail of every frame, so whether a game frame is requested or the status of the everdrive, the last two bytes would be the same.